Add option for invert mass window cut and Delta R for HLTMuonDimuonL3Filter

arsahasransu opened the pull request:
#### PR description:

  - This PR is a development towards CMS Run3 HLT where we trigger events with soft muons in them. To get a reasonable rate, new kinematic constraints on the dimuon events need to be placed at the HLT.
  - Added options to make a selection on the minimum angle required between L3 muons in the trigger
  - Added option to invert the mass window cut on the invariant mass of L3 muons 
  - This is based on the presentation at https://indico.cern.ch/event/1090283/#33-level-3-paths-for-singlet-t

  - Sam James Harper noticed that error management in the module can be improved.
  - It is ideal to throw an exception at the constructor level when that error renders the module of no use. This enable catching the error in the build phase instead of having to wait for a message from cmsRun.

#### PR validation:

<!-- Please replace this text with a description of which tests have been performed to verify the correctness of the PR, including the eventual addition of new code for testing like unit tests, test configurations, additions or updates to the runTheMatrix test workflows -->

  - Standard set of checks performed at https://cms-sw.github.io/PRWorkflow.html were done and passed.
  - A trigger was made with the suggested changes. We observed expected behavior from this trigger of interest.

jeongeun committed:
replace EDProducer with clone from cfipython

mseidel42 committed:
MahiDebugger Run 3 config

swagata87 committed:
new rho producer for HLT

swagata87 committed:
code cleanup

swagata87 committed:
added supporting text

swagata87 committed:
addressed review comments

swagata87 committed:
Addressed comments from Vincenzo and Joosep

swagata87 committed:
code-check

swagata87 committed:
Address comments from Marino and Vincenzo

swagata87 committed:
code check

SohamBhattacharya committed:
First commit.

SohamBhattacharya committed:
Code checks and formats.

SohamBhattacharya committed:
Turned on the trackster iteration filter flag.

swagata87 committed:
remove unnecessary headers

mseidel42 committed:
Apply Slava's suggestions, fix plot for peak in BX+1

mmusich committed:
remove redundant line causing compilation warning

swagata87 committed:
add packages to BuildFile

slomeo committed:
Restored the ShieldingME4 and fixed all the overlaps present inside the Muon System for 2026D86 Geometry

slomeo committed:
Restored cmsExtendedGeometry2026D86.xml

Commit:
Use ESGetToken in RecoHI/HiTracking

Commit:
Code check

arsahasransu committed:
Added inverse mass window selection for dimuon invariant mass and minimum angle between the muons selection

mseidel42 committed:
Move cfi to python/test

arsahasransu committed:
Fixed HLTMuonDimuonL3Filter to throw exception during construction

arsahasransu committed:
Merge branch 'master' into InverseDiMuonMassFilter_1210pre5

arsahasransu committed:
Merge pull request #1 from arsahasransu/InverseDiMuonMassFilter_1210pre5

Add option for invert mass window cut and Delta R for HLTMuonDimuonL3Filter

smuzaffar committed:
set NO_TEST_PREFIX flag for ASAN IB

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36250/26903

- This PR adds an extra 24KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36250/26903/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36250/26903/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

missirol commented:
@arsahasransu 

This PR can't proceed without first passing code checks. Please follow
https://cms-sw.github.io/PRWorkflow.html

Also, once you have fixed those, please squash all commits into 1 for clarity.

arsahasransu committed:
Fixed failed code checks from CMSSW PR

arsahasransu committed:
Merge pull request #2 from arsahasransu/InverseDiMuonMassFilter_1210pre5

Fixed failed code checks from CMSSW PR

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36250/26909

- This PR adds an extra 24KB to repository

cmsbuild commented:
A new Pull Request was created by  @arsahasransu (A.R.Sahasransu) for master.

It involves the following packages:

- HLTrigger/Muon (**hlt**)


@cmsbuild, @missirol, @Martin-Grunewald can you please review it and eventually sign? Thanks.
@Fedespring, @HuguesBrun, @calderona, @silviodonato, @abbiendi, @jhgoh, @Martin-Grunewald, @missirol, @sscruz, @CeliaFernandez, @trocino, @cericeci this is something you requested to watch as well.
@perrotta, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

missirol commented on HLTrigger/Muon/plugins/HLTMuonDimuonL3Filter.cc:
```diff
@@ -595,17 +602,21 @@ bool HLTMuonDimuonL3Filter::applyDiMuonSelection(const RecoChargedCandidateRef&
   LogDebug("HLTMuonDimuonL3Filter") << " ... 1-2 pt12= " << pt12;
 
+  // Angle between the muons
+  if (deltaR(p1, p2) < min_DiMuAngle_)
```
```suggestion
  if (applyDiMuonDeltaRCut_ and reco::deltaR2(p1, p2) < min_DiMuonDeltaR2_)
```

arsahasransu replied:
The default value of the min_DiMuonDeltaR2 ensures when someone doesn't want the deltaR cut, they do not need to change the value or assign a value to this value. Adding another parameter to this list doesn't add any extra function. This is the same philosophy as the other cuts placed in this module. Applying the new parameter also adds to the time required to run this module. Please correct me if I am missing something.

missirol replied:
> Adding another parameter to this list doesn't add any extra function.

It improves the readability of the code, clarifying when the DeltaR cut will be applied. Also, if someone uses for example `"MinDiMuonDeltaR" = cms.double(-1.)` in the python configuration, he/she would rightfully expect that the DeltaR cut won't be applied (this is how other modules usually behave), but then `min_DiMuonDeltaR2_` would be `== -1*-1 == 1`, and without `applyDiMuonDeltaRCut_ and [..]` one would just end up cutting on `DR2 < 1` unintentionally.

> Applying the new parameter also adds to the time required to run this module.

Not that I have done any measurements, but I doubt that `true and "calculationX"` is significantly slower than just `"calculationX"`. In fact, it's generally the opposite; in the majority of cases, modules will have "applyDiMuonDeltaRCut_ == false", and in that case `deltaR2` won't be calculated in the first place, saving time (however small this effect is).

arsahasransu replied:
I understand now. Thank you for the review. I will accept the commits relevant to the DeltaR2 cut.

missirol commented on HLTrigger/Muon/plugins/HLTMuonDimuonL3Filter.h:
```diff
@@ -59,29 +59,31 @@ class HLTMuonDimuonL3Filter : public HLTFilter {
-  double max_YPair_;                 // |rapidity| of pair
-  bool cutCowboys_;                  ///< if true, reject muon-track pairs that bend towards each other
-  const edm::InputTag theL3LinksLabel;                                //Needed to iterL3
+  bool fast_Accept_;  // flag to save time: stop processing after identification of the first valid pair
+  int min_N_;         // minimum number of muons to fire the trigger
+  double max_Eta_;    // Eta cut
+  int min_Nhits_;     // threshold on number of hits on muon
+  double max_Dr_;     // impact parameter cut
+  double max_Dz_;     // dz cut
+  int chargeOpt_;     // Charge option (0:nothing; +1:same charge, -1:opposite charge)
```
```suggestion
  const bool fast_Accept_;  // flag to save time: stop processing after identification of the first valid pair
  const int min_N_;         // minimum number of muons to fire the trigger
  const double max_Eta_;    // Eta cut
  const int min_Nhits_;     // threshold on number of hits on muon
  const double max_Dr_;     // impact parameter cut
  const double max_Dz_;     // dz cut
  const int chargeOpt_;     // Charge option (0:nothing; +1:same charge, -1:opposite charge)
```

missirol commented on HLTrigger/Muon/plugins/HLTMuonDimuonL3Filter.h:
```diff
@@ -59,29 +59,31 @@ class HLTMuonDimuonL3Filter : public HLTFilter {
+  const std::vector<double> min_InvMass_;             // minimum invariant mass of pair
+  const std::vector<double> max_InvMass_;             // maximum invariant mass of pair
+  const std::vector<int> invertDiMuonMassSelection_;  // condition to invert the invariant mass selection
+  double min_DiMuAngle_;                              // minimum angular difference between the muons
+  double min_Acop_;                                   // minimum acoplanarity
+  double max_Acop_;                                   // maximum acoplanarity
+  double min_PtBalance_;                              // minimum Pt difference
+  double max_PtBalance_;                              // maximum Pt difference
+  double nsigma_Pt_;                                  // pt uncertainty margin (in number of sigmas)
+  double max_DCAMuMu_;                                // DCA between the two muons
+  double max_YPair_;                                  // |rapidity| of pair
+  bool cutCowboys_;                                   ///< if true, reject muon-track pairs that bend towards each other
```
```suggestion
  const bool applyMinDiMuonDeltaR2Cut_;                     // apply cut on minimum Delta-R^2 distance between the muons
  const double min_DiMuDeltaR2_;                            // minimum Delta-R^2 distance between the muons
  const double min_Acop_;                                   // minimum acoplanarity
  const double max_Acop_;                                   // maximum acoplanarity
  const double min_PtBalance_;                              // minimum Pt difference
  const double max_PtBalance_;                              // maximum Pt difference
  const double nsigma_Pt_;                                  // pt uncertainty margin (in number of sigmas)
  const double max_DCAMuMu_;                                // DCA between the two muons
  const double max_YPair_;                                  // |rapidity| of pair
  const bool cutCowboys_;                                   ///< if true, reject muon-track pairs that bend towards each other
```

missirol commented on HLTrigger/Muon/plugins/HLTMuonDimuonL3Filter.cc:
```diff
@@ -131,13 +159,17 @@ void HLTMuonDimuonL3Filter::fillDescriptions(edm::ConfigurationDescriptions& des
   desc.add<vector<double> >("MinInvMass", v6);
   desc.add<vector<double> >("MaxInvMass", v7);
+  desc.add<vector<int> >("invertDiMuonMassSelection", v8);
+  desc.add<double>("MinDiMuAngle", 0.0);
```
```suggestion
  desc.add<double>("MinDiMuonDeltaR", -1.);
```

missirol commented on HLTrigger/Muon/plugins/HLTMuonDimuonL3Filter.cc:
```diff
@@ -75,6 +75,8 @@ HLTMuonDimuonL3Filter::HLTMuonDimuonL3Filter(const edm::ParameterSet& iConfig)
       max_PtMin_(iConfig.getParameter<vector<double> >("MaxPtMin")),
       min_InvMass_(iConfig.getParameter<vector<double> >("MinInvMass")),
       max_InvMass_(iConfig.getParameter<vector<double> >("MaxInvMass")),
+      invertDiMuonMassSelection_(iConfig.getParameter<vector<int> >("invertDiMuonMassSelection")),
+      min_DiMuAngle_(iConfig.getParameter<double>("MinDiMuAngle")),
```
```suggestion
      applyMinDiMuonDeltaR2Cut_(iConfig.getParameter<double>("MinDiMuonDeltaR") > 0.),
      min_DiMuonDeltaR2_(iConfig.getParameter<double>("MinDiMuonDeltaR") * iConfig.getParameter<double>("MinDiMuonDeltaR")),
```

missirol commented on HLTrigger/Muon/plugins/HLTMuonDimuonL3Filter.cc:
```diff
@@ -88,14 +90,40 @@ HLTMuonDimuonL3Filter::HLTMuonDimuonL3Filter(const edm::ParameterSet& iConfig)
-                                    << Out(min_InvMass_) << " " << Out(max_InvMass_) << " " << min_Acop_ << " "
-                                    << max_Acop_ << " " << min_PtBalance_ << " " << max_PtBalance_ << " " << nsigma_Pt_
-                                    << " " << max_DCAMuMu_ << " " << max_YPair_;
+                                    << Out(min_InvMass_) << " " << Out(max_InvMass_) << " " << min_DiMuAngle_ << " "
```
```suggestion
                                    << Out(min_InvMass_) << " " << Out(max_InvMass_) << " " << applyMinDiMuonDeltaR2Cut_ << " " << math::sqrt(min_DiMuonDeltaR2_) << " "
```

missirol commented on HLTrigger/Muon/plugins/HLTMuonDimuonL3Filter.cc:
```diff
@@ -131,13 +159,17 @@ void HLTMuonDimuonL3Filter::fillDescriptions(edm::ConfigurationDescriptions& des
   v6.push_back(2.8);
   vector<double> v7;
   v7.push_back(3.4);
+  vector<int> v8;
+  v8.push_back(0);
```
```suggestion
```

missirol commented on HLTrigger/Muon/plugins/HLTMuonDimuonL3Filter.cc:
```diff
@@ -131,13 +159,17 @@ void HLTMuonDimuonL3Filter::fillDescriptions(edm::ConfigurationDescriptions& des
   desc.add<vector<double> >("MaxPtMin", v5);
   desc.add<vector<double> >("MinInvMass", v6);
   desc.add<vector<double> >("MaxInvMass", v7);
+  desc.add<vector<int> >("invertDiMuonMassSelection", v8);
```
```suggestion
```

missirol commented on HLTrigger/Muon/plugins/HLTMuonDimuonL3Filter.cc:
```diff
@@ -595,17 +602,21 @@ bool HLTMuonDimuonL3Filter::applyDiMuonSelection(const RecoChargedCandidateRef&
-      continue;
-    if (invmass > max_InvMass_[iv])
-      continue;
+    if ((invertDiMuonMassSelection_[iv] != 0) && invmass > min_InvMass_[iv] && invmass < max_InvMass_[iv])
+      return false;
+    if ((invertDiMuonMassSelection_[iv] == 0) && (invmass < min_InvMass_[iv] || invmass > max_InvMass_[iv]))
+      return false;
```
```suggestion
    if (invmass < min_InvMass_[iv])
      continue;
    if (invmass > max_InvMass_[iv])
      continue;
```

missirol commented on HLTrigger/Muon/plugins/HLTMuonDimuonL3Filter.h:
```diff
@@ -59,29 +59,31 @@ class HLTMuonDimuonL3Filter : public HLTFilter {
+  const std::vector<double> max_PtMin_;               // maximum Pt for muon with min Pt in pair
+  const std::vector<double> min_InvMass_;             // minimum invariant mass of pair
+  const std::vector<double> max_InvMass_;             // maximum invariant mass of pair
+  const std::vector<int> invertDiMuonMassSelection_;  // condition to invert the invariant mass selection
```
```suggestion
```

arsahasransu replied:
In principle, I could choose two mass windows, and then I do not need to add the invert mass window selection choice. However, this increases the number of conditions checked from 19 to 28, for every event with a muon pair because now we have to check against a pair of values for each of the following vectors.
  - min_PtMin_
  - min_PtMax_
  - max_PtMin_
  - min_PtPair_
  - max_PtPair_
 
This can potentially cause a significant increase in the timing of an event. Please let me know what are your thoughts on this.

missirol replied:
> This can potentially cause a significant increase in the timing of an event. Please let me know what are your thoughts on this.

This would have to be quantified (although I would be surprised if this specific filter, even in its possible new configurations, would add significant time to the overall HLT reconstruction). Assuming there will be a new path which uses a new configuration for this plugin (inverted mass-windows), is the timing of that path significant, and is this filter a significant contributor to that? (I'm not asking now, but these are the questions, imho). Readibility of the code, and making sure the default behaviour of the plugin is preserved upon the introduction of new optional features are also important factors.

arsahasransu replied:
The expected rate of the HLT that this path will be used in is ~5Hz. This filter is the last step in this path. Dimuon triggers typically take less time to run at HLT. Do you think I should make a time study regarding this? Or is there any way we can improve the readability of the code with the added option to invertDiMuonMass and move forward?

missirol replied:
Based on this info, the easiest solution in my opinion is to not modify the cpp implementation of the plugin (which is already not particularly easy to read..), and to simply configure it differently at the python level. If later on it turns out that there is creates a significant timing increase (before the new path is integrated), then we'll have to think about improving the plugin.

By the way, if `invertDiMuonMass` is not removed, the PR as is will not pass the tests. The menus would have to be customised appropriately in order for the relevant modules to use the correct values of the new `vint` (a list of zeros, I guess, but with different size for different modules). This change is not terribly complex, but it is missing nonetheless at the moment.

missirol reviewed (commented):
Here's the first round of comments.

 - I don't see the necessity to introduce the `invertDiMuonMassSelection` parameter; couldn't the same be obtained through a proper configuration of the existing plugin? e.g.
 ```
    MinPtPair = cms.vdouble( 0., 0. ),
    MaxPtPair = cms.vdouble( 1.0E125, 1.0E125 ),
    MinPtMax = cms.vdouble( 0.0, 0.0 ),
    MinPtMin = cms.vdouble( 0.0, 0.0 ),
    MaxPtMin = cms.vdouble( 1.0E125, 1.0E125 ),
    MinInvMass = cms.vdouble( -1.0, 98. ),
    MaxInvMass = cms.vdouble( 84., 1.0E125 ),
 ```
  - the DeltaR cut should be disabled by default, and converted internally to a cut on DeltaR^2 for better computational performance.

arsahasransu committed:
Changed the exception throw message HLTrigger/Muon/plugins/HLTMuonDimuonL3Filter.cc

Co-authored-by: Marino Missiroli <m.missiroli@cern.ch>

arsahasransu committed:
Class variables defined as const. HLTrigger/Muon/plugins/HLTMuonDimuonL3Filter.h

Co-authored-by: Marino Missiroli <m.missiroli@cern.ch>

Adding info. to the debug log of HLTrigger/Muon/plugins/HLTMuonDimuonL3Filter.cc

Co-authored-by: Marino Missiroli <m.missiroli@cern.ch>

Declare class variables as constant HLTrigger/Muon/plugins/HLTMuonDimuonL3Filter.h

Co-authored-by: Marino Missiroli <m.missiroli@cern.ch>

Update debug log HLTrigger/Muon/plugins/HLTMuonDimuonL3Filter.cc

Co-authored-by: Marino Missiroli <m.missiroli@cern.ch>

Fixed failed code checks from CMSSW PR

Update HLTrigger/Muon/plugins/HLTMuonDimuonL3Filter.cc

Co-authored-by: Marino Missiroli <m.missiroli@cern.ch>

Update HLTrigger/Muon/plugins/HLTMuonDimuonL3Filter.cc

Co-authored-by: Marino Missiroli <m.missiroli@cern.ch>

Update HLTrigger/Muon/plugins/HLTMuonDimuonL3Filter.cc

Co-authored-by: Marino Missiroli <m.missiroli@cern.ch>

Update HLTrigger/Muon/plugins/HLTMuonDimuonL3Filter.cc

Co-authored-by: Marino Missiroli <m.missiroli@cern.ch>

Update HLTrigger/Muon/plugins/HLTMuonDimuonL3Filter.cc

Co-authored-by: Marino Missiroli <m.missiroli@cern.ch>

Fixed failed code checks from CMSSW PR

Fixed failed code checks from CMSSW PR

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36250/26915

- This PR adds an extra 20KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36250/26915/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36250/26915/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36250/26918

ERROR: Build errors found during clang-tidy run.
```
HLTrigger/Muon/plugins/HLTMuonDimuonL3Filter.cc:79:7: error: member initializer 'min_DiMuAngle_' does not name a non-static data member or base class [clang-diagnostic-error]
      min_DiMuAngle_(iConfig.getParameter<double>("MinDiMuAngle")),
      ^
HLTrigger/Muon/plugins/HLTMuonDimuonL3Filter.cc:118:138: error: no member named 'sqrt' in namespace 'math' [clang-diagnostic-error]
                                    << Out(min_InvMass_) << " " << Out(max_InvMass_) << " " << applyMinDiMuonDeltaR2Cut_ << " " << math::sqrt(min_DiMuonDeltaR2_) << " "
                                                                                                                                         ^
HLTrigger/Muon/plugins/HLTMuonDimuonL3Filter.cc:118:143: error: use of undeclared identifier 'min_DiMuonDeltaR2_' [clang-diagnostic-error]
                                    << Out(min_InvMass_) << " " << Out(max_InvMass_) << " " << applyMinDiMuonDeltaR2Cut_ << " " << math::sqrt(min_DiMuonDeltaR2_) << " "
                                                                                                                                              ^
HLTrigger/Muon/plugins/HLTMuonDimuonL3Filter.cc:600:24: error: use of undeclared identifier 'min_DiMuAngle_' [clang-diagnostic-error]
  if (deltaR(p1, p2) < min_DiMuAngle_)
                       ^
Suppressed 1520 warnings (1519 in non-user code, 1 with check filters).
--
gmake: *** [config/SCRAM/GMake/Makefile.coderules:129: code-checks] Error 2
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36250/26920

ERROR: Build errors found during clang-tidy run.
```
HLTrigger/Muon/plugins/HLTMuonDimuonL3Filter.cc:79:7: error: member initializer 'min_DiMuAngle_' does not name a non-static data member or base class [clang-diagnostic-error]
      min_DiMuAngle_(iConfig.getParameter<double>("MinDiMuAngle")),
      ^
HLTrigger/Muon/plugins/HLTMuonDimuonL3Filter.cc:124:138: error: no member named 'sqrt' in namespace 'math' [clang-diagnostic-error]
                                    << Out(min_InvMass_) << " " << Out(max_InvMass_) << " " << applyMinDiMuonDeltaR2Cut_ << " " << math::sqrt(min_DiMuonDeltaR2_) << " "
                                                                                                                                         ^
HLTrigger/Muon/plugins/HLTMuonDimuonL3Filter.cc:124:143: error: use of undeclared identifier 'min_DiMuonDeltaR2_' [clang-diagnostic-error]
                                    << Out(min_InvMass_) << " " << Out(max_InvMass_) << " " << applyMinDiMuonDeltaR2Cut_ << " " << math::sqrt(min_DiMuonDeltaR2_) << " "
                                                                                                                                              ^
HLTrigger/Muon/plugins/HLTMuonDimuonL3Filter.cc:606:24: error: use of undeclared identifier 'min_DiMuAngle_' [clang-diagnostic-error]
  if (deltaR(p1, p2) < min_DiMuAngle_)
                       ^
Suppressed 1520 warnings (1519 in non-user code, 1 with check filters).
--
gmake: *** [config/SCRAM/GMake/Makefile.coderules:129: code-checks] Error 2
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36250/26923

ERROR: Build errors found during clang-tidy run.
```
HLTrigger/Muon/plugins/HLTMuonDimuonL3Filter.cc:79:7: error: initializer 'min_DiMuonDeltaR2_' does not name a non-static data member or base class; did you mean the member 'min_DiMuDeltaR2_'? [clang-diagnostic-error]
      min_DiMuonDeltaR2_(iConfig.getParameter<double>("MinDiMuonDeltaR") * iConfig.getParameter<double>("MinDiMuonDeltaR")),
      ^~~~~~~~~~~~~~~~~~
      min_DiMuDeltaR2_
--
HLTrigger/Muon/plugins/HLTMuonDimuonL3Filter.cc:124:138: error: no member named 'sqrt' in namespace 'math' [clang-diagnostic-error]
                                    << Out(min_InvMass_) << " " << Out(max_InvMass_) << " " << applyMinDiMuonDeltaR2Cut_ << " " << math::sqrt(min_DiMuonDeltaR2_) << " "
                                                                                                                                         ^
HLTrigger/Muon/plugins/HLTMuonDimuonL3Filter.cc:124:143: error: use of undeclared identifier 'min_DiMuonDeltaR2_' [clang-diagnostic-error]
                                    << Out(min_InvMass_) << " " << Out(max_InvMass_) << " " << applyMinDiMuonDeltaR2Cut_ << " " << math::sqrt(min_DiMuonDeltaR2_) << " "
                                                                                                                                              ^
HLTrigger/Muon/plugins/HLTMuonDimuonL3Filter.cc:603:24: error: use of undeclared identifier 'min_DiMuAngle_' [clang-diagnostic-error]
  if (deltaR(p1, p2) < min_DiMuAngle_)
                       ^
Suppressed 1516 warnings (1515 in non-user code, 1 with check filters).
--
gmake: *** [config/SCRAM/GMake/Makefile.coderules:129: code-checks] Error 2
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36250/26925

- This PR adds an extra 32KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36250/26925/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36250/26925/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36250/26933

- This PR adds an extra 32KB to repository

cmsbuild commented:
Pull request #36250 was updated. @cmsbuild, @missirol, @Martin-Grunewald can you please check and sign again.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36250/26934

- This PR adds an extra 80KB to repository

cmsbuild commented:
Pull request #36250 was updated. @cmsbuild, @missirol, @Martin-Grunewald can you please check and sign again.

missirol commented:
Thanks @arsahasransu, the changes now look okay.

> Also, once you have fixed those, please squash all commits into 1 for clarity.

As mentioned [above](https://github.com/cms-sw/cmssw/pull/36250#issuecomment-979384087), I kindly ask you to squash the commits done so far into 1 single commit. Right now I see 20 commits, most of which are "merge" commits, code-checks fixes, and integration of single review comments; this history does not need to be retained. You can find instructions for squashing commits online, and you will then have to force-push (`push -f`) your `master` branch, such that the PR will contain 1 commit with all changes done so far. After that, we can proceed.

arsahasransu commented:
I made a major mistake while trying to squash all the changes into a single commit. So I have closed this PR. I will make a fresh PR soon.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-36250/26935

- This PR adds an extra 72KB to repository

- Found files with invalid states:
  - RecoLocalCalo/HcalRecAlgos/python/mahiDebugger_cfi.py:
    - Added: 52478221f99d553e6c07a6073eb402e5d80eafe3
    - Modified: 835005939fba371c29037649744e9d731614e49a
    - Deleted: 35e561478e8905063825455c3be5fa9c18699595

cmsbuild commented:
Pull request #36250 was updated. @Martin-Grunewald, @civanch, @Dr15Jones, @makortel, @cvuosalo, @emanueleusai, @ianna, @mdhildreth, @cmsbuild, @missirol, @jfernan2, @ahmad3213, @slava77, @jpata, @pmandrik, @pbo0, @rvenditti can you please check and sign again.
